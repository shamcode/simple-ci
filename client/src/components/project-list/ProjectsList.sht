{% import path from 'sham-ui-router/lib/params' %}

<div class="project-list">
    {% for project of projects %}
        <a
            class="project-card"
            :hrefto={{path("project-detail").param("id", project.id)}}
        >
            <span class="name">
                {{project.name}}
            </span>
        </a>
    {% endfor %}
    <a
        class="project-card project-card-create"
        :hrefto={{path("project-create")}}
    >
        <div class="create-icon"></div>
    </a>
</div>
