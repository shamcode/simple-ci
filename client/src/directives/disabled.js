export default class {
    constructor() {
        this.node = null;
    }

    bind( node ) {
        this.node = node;
    }

    unbind() {
        this.node = null;
    }

    update( value ) {
        this.node.disabled = value;
    }
}
