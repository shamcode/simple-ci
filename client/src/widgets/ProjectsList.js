import { options, DI } from 'sham-ui';
import ProjectsListTemplate from './ProjectsList.sht';

export default class extends ProjectsListTemplate {
    get router() {
        return DI.resolve( 'router' );
    }

    @options get createProjectURL() {
        return this.router.generate( 'project-create', {} );
    }

    get createProjectLinkNode() {
        return this.querySelector( '.project-card-create' );
    }

    bindEvents() {
        this.__bindedClickToCreateNewProject = this._clickToCreateNewProject.bind( this );
        this.createProjectLinkNode.addEventListener(
            'click',
            this.__bindedClickToCreateNewProject
        )
    }

    destroy() {
        this.createProjectLinkNode.removeEventListener(
            'click',
            this.__bindedClickToCreateNewProject
        );
    }

    remove() {
        this.createProjectLinkNode.removeEventListener(
            'click',
            this.__bindedClickToCreateNewProject
        );
        super.remove( ...arguments );
    }

    _clickToCreateNewProject( e ) {
        e.preventDefault();
        this.router.navigate( this.createProjectLinkNode.getAttribute( 'href' ) );
    }
}