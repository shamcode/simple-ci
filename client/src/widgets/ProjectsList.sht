<div class="project-list">
    {% for project of projects %}
        <a class="project-card" href="#">
            <span class="name">
                {{project.name}}
            </span>
        </a>
    {% endfor %}
    <a class="project-card project-card-create" href="{{createProjectURL}}">
        <div class="create-icon"></div>
    </a>
</div>