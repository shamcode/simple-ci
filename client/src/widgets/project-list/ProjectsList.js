import { options, DI } from 'sham-ui';
import ProjectsListTemplate from './ProjectsList.sht';

export default class extends ProjectsListTemplate {
    get router() {
        return DI.resolve( 'router' );
    }

    @options get createProjectURL() {
        return this.router.generate( 'project-create', {} );
    }

    detailPageUrlForProject( id ) {
        return this.router.generate( 'project-detail', { id } );
    }

    get createProjectLinkNode() {
        return this.querySelector( '.project-card-create' );
    }

    _clickToDetailProject( e ) {
        e.preventDefault();
        this.router.navigate( e.currentTarget.getAttribute( 'href' ) );
    }

    _clickToCreateNewProject( e ) {
        e.preventDefault();
        this.router.navigate( this.createProjectLinkNode.getAttribute( 'href' ) );
    }
}