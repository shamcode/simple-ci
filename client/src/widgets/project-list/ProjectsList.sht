<div class="project-list">
    {% for project of projects %}
        <a
                class="project-card"
                href={{this.detailPageUrlForProject(project.id)}}
                :onclick={{this._clickToDetailProject.bind(this)}}
        >
            <span class="name">
                {{project.name}}
            </span>
        </a>
    {% endfor %}
    <a
            class="project-card project-card-create"
            href="{{createProjectURL}}"
            :onclick={{this._clickToCreateNewProject.bind(this)}}
    >
        <div class="create-icon"></div>
    </a>
</div>