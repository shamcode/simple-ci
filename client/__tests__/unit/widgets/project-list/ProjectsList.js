import ProjectsList from '../../../../src/widgets/project-list/ProjectsList';
import directives from '../../../../src/directives/event-listener';
import renderer from 'sham-ui-test-helpers';

afterEach( () => {
    DI.bind( 'router', null );
} );

it( 'renders correctly', () => {
    const generateMock = jest.fn().mockReturnValueOnce( '/' );
    DI.bind( 'router', {
        generate: generateMock
    } );

    const meta = renderer( ProjectsList, { directives } );
    expect( generateMock.mock.calls.length ).toBe( 1 );
    expect( generateMock.mock.calls[ 0 ][ 0 ] ).toEqual( "project-create" );
    expect( Object.keys( generateMock.mock.calls[ 0 ][ 1 ] ).length ).toBe( 0 );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'go to create page', () => {
    const navigateMock = jest.fn();
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValueOnce( '/' ),
        navigate: navigateMock
    } );

    const meta = renderer( ProjectsList, { directives } );
    meta.widget.querySelector( '.project-card-create' ).click();

    expect( navigateMock.mock.calls.length ).toBe( 1 );
    expect( meta.toJSON() ).toMatchSnapshot();
} );

it( 'renders list', () => {
    DI.bind( 'router', {
        generate: jest.fn().mockReturnValueOnce( '/' )
    } );

    const meta = renderer( ProjectsList, {
        directives,
        projects: [
            { id: 1, name: 'Simple CI', cwd: '/tmp/' },
            { id: 1, name: 'ShamUI', cwd: '/home/' }
        ]
    } );
    expect( meta.toJSON() ).toMatchSnapshot();
} );