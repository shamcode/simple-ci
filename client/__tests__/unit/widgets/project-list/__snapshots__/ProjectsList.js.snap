// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`go to create page 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<div class=\\"project-list\\">
  <!--for-->
  <a class=\\"project-card project-card-create\\" href=\\"/\\">
    <div class=\\"create-icon\\"></div>
  </a>
</div>
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "createProjectURL": "/",
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;

exports[`renders correctly 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
  },
  "html": "
<div class=\\"project-list\\">
  <!--for-->
  <a class=\\"project-card project-card-create\\" href=\\"/\\">
    <div class=\\"create-icon\\"></div>
  </a>
</div>
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "createProjectURL": "/",
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;

exports[`renders list 1`] = `
Object {
  "ID": "widget",
  "constructorOptions": Object {
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "projects": Array [
      Object {
        "cwd": "/tmp/",
        "id": 1,
        "name": "Simple CI",
      },
      Object {
        "cwd": "/home/",
        "id": 1,
        "name": "ShamUI",
      },
    ],
  },
  "html": "
<div class=\\"project-list\\">
  <a class=\\"project-card\\" href=\\"undefined\\"><span class=\\"name\\">Simple CI</span></a>
  <a class=\\"project-card\\" href=\\"undefined\\"><span class=\\"name\\">ShamUI</span></a>
  <!--for-->
  <a class=\\"project-card project-card-create\\" href=\\"/\\">
    <div class=\\"create-icon\\"></div>
  </a>
</div>
",
  "nodes": Array [
    Object {
      "ID": undefined,
      "constructorOptions": null,
      "html": null,
      "nodes": null,
      "options": null,
    },
  ],
  "options": Object {
    "actionSequence": Array [
      "render",
      "bindEvents",
    ],
    "createProjectURL": "/",
    "directives": Object {
      "EventListener": [Function],
      "disabled": [Function],
      "onclick": [Function],
      "oninput": [Function],
      "onload": [Function],
      "onresize": [Function],
      "onscroll": [Function],
      "onselect": [Function],
      "onsubmit": [Function],
    },
    "projects": Array [
      Object {
        "cwd": "/tmp/",
        "id": 1,
        "name": "Simple CI",
      },
      Object {
        "cwd": "/home/",
        "id": 1,
        "name": "ShamUI",
      },
    ],
    "renderAsync": false,
    "renderDependence": Array [],
    "types": Array [],
  },
}
`;
